ClassFanOutComplexity fails to count all classes in multicatch and implements clause

I have read check documentation: https://checkstyle.sourceforge.io/config_metrics.html#ClassFanOutComplexity
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```

➜  src javac Test.java
Note: Test.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.

➜  src cat config.xml 
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="ClassFanOutComplexity">
          <property name="max" value="1"/>
<!--          <property name="memberModifiers" value="LITERAL_PUBLIC, FINAL"/>-->
        </module>
    </module>
</module>

➜  src cat Test.java
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.EmptyStackException;

public class Test {
    public static void main(String[] args) {
        try {
            System.out.println(args[7]);
            File myFile = new File("myfile.txt"); // 1

            InputStream stream = myFile.toURL().openStream(); // 2
        } catch (IOException  | EmptyStackException e) { // 3,4

        }
    }

}

➜  src java -jar checkstyle-8.43-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/Desktop/Tester/src/Test.java:6:1: Class Fan-Out Complexity is 3 (max allowed is 1). [ClassFanOutComplexity]
Audit done.
Checkstyle ends with 1 errors.
➜  src 


```

I would expect the "Class Fan-Out Complexity" to be 4.

--- a/config/checkstyle_checks.xml
+++ b/config/checkstyle_checks.xml
@@ -660,7 +660,8 @@
                        FileNotFoundException, TestException, Log, Sets, Multimap,
                        TokenStreamRecognitionException, RecognitionException,
                        TokenStreamException, IOException, Override, Deprecated, SafeVarargs,
-                       SuppressWarnings, FunctionalInterface"/>
+                       SuppressWarnings, FunctionalInterface, NoSuchMethodException,
+                       InvocationTargetException, IllegalAccessException"/>
     </module>
     <module name="CyclomaticComplexity">
       <property name="switchBlockAsSingleDecisionPoint" value="true"/>

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/AbstractClassCouplingCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/AbstractClassCouplingCheck.java
@@ -38,7 +38,6 @@ import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
 import com.puppycrawl.tools.checkstyle.api.DetailAST;
 import com.puppycrawl.tools.checkstyle.api.FullIdent;
 import com.puppycrawl.tools.checkstyle.api.TokenTypes;
-import com.puppycrawl.tools.checkstyle.utils.CheckUtil;
 import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
 import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
 
@@ -376,10 +375,16 @@ public abstract class AbstractClassCouplingCheck extends AbstractCheck {
          * @param ast type to process.
          */
         public void visitType(DetailAST ast) {
-            final String fullTypeName = CheckUtil.createFullType(ast).getText();
-            final String trimmed = BRACKET_PATTERN
-                    .matcher(fullTypeName).replaceAll("");
-            addReferencedClassName(trimmed);
+            DetailAST child = ast.getFirstChild();
+            while (child != null) {
+                if (TokenUtil.isOfType(child, TokenTypes.IDENT, TokenTypes.DOT)) {
+                    final String fullTypeName = FullIdent.createFullIdent(child).getText();
+                    final String trimmed = BRACKET_PATTERN
+                            .matcher(fullTypeName).replaceAll("");
+                    addReferencedClassName(trimmed);
+                }
+                child = child.getNextSibling();
+            }
         }
 
         /**

--- a/src/main/java/com/puppycrawl/tools/checkstyle/utils/CheckUtil.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/utils/CheckUtil.java
@@ -28,7 +28,6 @@ import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
 import com.puppycrawl.tools.checkstyle.api.DetailAST;
-import com.puppycrawl.tools.checkstyle.api.FullIdent;
 import com.puppycrawl.tools.checkstyle.api.TokenTypes;
 import com.puppycrawl.tools.checkstyle.checks.naming.AccessModifierOption;
 
@@ -74,16 +73,6 @@ public final class CheckUtil {
 
     /** Prevent instances. */
     private CheckUtil() {
-    }
-
-    /**
-     * Creates {@code FullIdent} for given type node.
-     *
-     * @param typeAST a type node.
-     * @return {@code FullIdent} for given type.
-     */
-    public static FullIdent createFullType(final DetailAST typeAST) {
-        return FullIdent.createFullIdent(typeAST.getFirstChild());
     }
 
     /**

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheckTest.java
@@ -232,6 +232,15 @@ public class ClassFanOutComplexityCheckTest extends AbstractModuleTestSupport {
     }
 
     @Test
+    public void testImplementsAndNestedCount() throws Exception {
+        final String[] expected = {
+            "26:1: " + getCheckMessage(MSG_KEY, 3, 0),
+        };
+        verifyWithInlineConfigParser(
+                getPath("InputClassFanOutComplexityImplementsAndNestedCount.java"), expected);
+    }
+
+    @Test
     public void testClassFanOutComplexityRecords() throws Exception {
         final String[] expected = {
             "32:1: " + getCheckMessage(MSG_KEY, 4, 2),
@@ -263,10 +272,13 @@ public class ClassFanOutComplexityCheckTest extends AbstractModuleTestSupport {
     }
 
     @Test
-    public void testClassFanOutComplexityRemoveMultiCatchBitwiseOr() throws Exception {
-        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+    public void testClassFanOutComplexityMultiCatchBitwiseOr() throws Exception {
+        final String[] expected = {
+            "27:1: " + getCheckMessage(MSG_KEY, 5, 4),
+        };
+
         verifyWithInlineConfigParser(
-                getPath("InputClassFanOutComplexityRemoveMultiCatchBitwiseOr.java"), expected);
+                getPath("InputClassFanOutComplexityMultiCatchBitwiseOr.java"), expected);
     }
 
     /**

--- a/src/test/java/com/puppycrawl/tools/checkstyle/utils/CheckUtilTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/utils/CheckUtilTest.java
@@ -23,7 +23,6 @@ import static com.puppycrawl.tools.checkstyle.internal.utils.TestUtil.findTokenI
 import static com.puppycrawl.tools.checkstyle.internal.utils.TestUtil.isUtilsClassHasPrivateConstructor;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertFalse;
-import static org.junit.jupiter.api.Assertions.assertNull;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 import static org.junit.jupiter.api.Assertions.fail;
 
@@ -40,7 +39,6 @@ import com.puppycrawl.tools.checkstyle.AbstractPathTestSupport;
 import com.puppycrawl.tools.checkstyle.DetailAstImpl;
 import com.puppycrawl.tools.checkstyle.JavaParser;
 import com.puppycrawl.tools.checkstyle.api.DetailAST;
-import com.puppycrawl.tools.checkstyle.api.FullIdent;
 import com.puppycrawl.tools.checkstyle.api.TokenTypes;
 import com.puppycrawl.tools.checkstyle.checks.naming.AccessModifierOption;
 
@@ -153,23 +151,6 @@ public class CheckUtilTest extends AbstractPathTestSupport {
             final String actualExceptionMsg = exc.getMessage();
             assertEquals(expectedExceptionMsg, actualExceptionMsg, "Invalid exception message");
         }
-    }
-
-    @Test
-    public void testCreateFullType() throws Exception {
-        final DetailAST typeNode = getNodeFromFile(TokenTypes.TYPE);
-
-        assertEquals("Map[13x12]", CheckUtil.createFullType(typeNode).toString(),
-                "Invalid full type");
-    }
-
-    @Test
-    public void testCreateFullTypeOfArray() throws Exception {
-        final DetailAST arrayTypeNode = getNodeFromFile(TokenTypes.VARIABLE_DEF)
-                .getNextSibling().getFirstChild().getNextSibling();
-
-        assertEquals("int[][14x14]", CheckUtil.createFullType(arrayTypeNode).toString(),
-                "Invalid full type");
     }
 
     @Test
@@ -397,13 +378,6 @@ public class CheckUtilTest extends AbstractPathTestSupport {
         expected.add("my.Class");
         expected.add("Class");
         assertEquals(expected, actual, "Result is not expected");
-    }
-
-    @Test
-    public void testNullChild() {
-        final DetailAstImpl root = new DetailAstImpl();
-        final FullIdent ident = CheckUtil.createFullType(root);
-        assertNull(ident.getDetailAst(), "'ident' should be null");
     }
 
     private DetailAST getNodeFromFile(int type) throws Exception {

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexityImplementsAndNestedCount.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexityImplementsAndNestedCount.java
@@ -1,6 +1,6 @@
 /*
 ClassFanOutComplexity
-max = 4
+max = 0
 excludedClasses = (default)ArrayIndexOutOfBoundsException, ArrayList, Boolean, Byte, \
                   Character, Class, Collection, Deprecated, Deque, Double, DoubleStream, \
                   EnumSet, Exception, Float, FunctionalInterface, HashMap, HashSet, \
@@ -20,32 +20,14 @@ excludedPackages = (default)
 
 package com.puppycrawl.tools.checkstyle.checks.metrics.classfanoutcomplexity;
 
-import java.io.File;
-import java.io.IOException;
-import java.io.InputStream;
+import java.io.Serializable;
+import java.util.RandomAccess;
 
-public class InputClassFanOutComplexityRemoveMultiCatchBitwiseOr { // ok
-    public static void main(String[] args) {
-        try {
-            System.out.println(args[7]);
-            File myFile = new File("myfile.txt");
-            InputStream stream = myFile.toURL().openStream();
-            foo1();
-        } catch (ArrayIndexOutOfBoundsException | IOException
-                | SQLException | SecurityException | OneMoreException e) {
+public class InputClassFanOutComplexityImplementsAndNestedCount // violation '.* Complexity is 3 .*'
+    extends Outer.Nested implements Serializable, RandomAccess {
 
-        }
-    }
+}
 
-    private static void foo1() throws RuntimeException, SQLException, OneMoreException {
-
-    }
-
-    private class SQLException extends Exception {
-
-    }
-
-    private class OneMoreException extends Exception {
-
-    }
+class Outer {
+    static class Nested {}
 }

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexityMultiCatchBitwiseOr.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/classfanoutcomplexity/InputClassFanOutComplexityMultiCatchBitwiseOr.java
@@ -1,0 +1,51 @@
+/*
+ClassFanOutComplexity
+max = 4
+excludedClasses = (default)ArrayIndexOutOfBoundsException, ArrayList, Boolean, Byte, \
+                  Character, Class, Collection, Deprecated, Deque, Double, DoubleStream, \
+                  EnumSet, Exception, Float, FunctionalInterface, HashMap, HashSet, \
+                  IllegalArgumentException, IllegalStateException, IndexOutOfBoundsException, \
+                  IntStream, Integer, LinkedHashMap, LinkedHashSet, LinkedList, List, Long, \
+                  LongStream, Map, NullPointerException, Object, Optional, OptionalDouble, \
+                  OptionalInt, OptionalLong, Override, Queue, RuntimeException, SafeVarargs, \
+                  SecurityException, Set, Short, SortedMap, SortedSet, Stream, String, \
+                  StringBuffer, StringBuilder, SuppressWarnings, Throwable, TreeMap, TreeSet, \
+                  UnsupportedOperationException, Void, boolean, byte, char, double, float, \
+                  int, long, short, var, void
+excludeClassesRegexps = (default)^$
+excludedPackages = (default)
+
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.metrics.classfanoutcomplexity;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+
+public class InputClassFanOutComplexityMultiCatchBitwiseOr { // violation '.* Complexity is 5 .*'
+    public static void main(String[] args) {
+        try {
+            System.out.println(args[7]);
+            File myFile = new File("myfile.txt");
+            InputStream stream = myFile.toURL().openStream();
+            foo1();
+        } catch (ArrayIndexOutOfBoundsException | IOException
+                | SQLException | SecurityException | OneMoreException e) {
+
+        }
+    }
+
+    private static void foo1() throws RuntimeException, SQLException, OneMoreException {
+
+    }
+
+    private class SQLException extends Exception {
+
+    }
+
+    private class OneMoreException extends Exception {
+
+    }
+}

